### PR TITLE
Add flatSubmit method to EventLoop (#426)

### PR DESCRIPTION
**Pull request by 2bjake**
Add flatSubmit method to EventLoop (#426)

### Motivation:

Calling EventLoop's submit method with a task that returns an EventLoopFuture results in a double future. flatSubmit provides an alternative which does not return a double future.

### Modifications:

Added a flatSubmit() default implementation that returns the result of calling flatMap { $0 } on the EventLoopFuture returned from task.

### Result:

Additive change only. flatSubmit method on EventLoop will be available for use.

### CODE REVIEW

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

File: Sources/NIO/EventLoop.swift [hunk 1]
```diff
@@ -232,6 +232,13 @@ public protocol EventLoop: EventLoopGroup {
     /// - returns: `EventLoopFuture` that is notified once the task was executed.
     func submit<T>(_ task: @escaping () throws -> T) -> EventLoopFuture<T>
 
+    /// Submit a given task to be executed by the `EventLoop`. Once the execution is complete the returned `EventLoopFuture` is notified.
+    ///
+    /// - parameters:
+    ///     - task: The closure that will be submitted to the `EventLoop` for execution.
+    /// - returns: `EventLoopFuture` that is notified once the task was executed.
+    func flatSubmit<T>(_ task: @escaping () throws -> EventLoopFuture<T>) -> EventLoopFuture<T>
```
Comment (weissi) [inline Sources/NIO/EventLoop.swift]: We don't need this as a customisation point on the protocol, you can just remove this line and provide an implementation in the extension (as you already do).

File: Sources/NIO/EventLoop.swift [hunk 2]
```diff
@@ -465,6 +472,18 @@ extension EventLoop {
...
 
+    /// Submit `task` to be run on this `EventLoop`.
+    ///
+    /// The returned `EventLoopFuture` will be completed when `task` has finished running. It will be identical to
+    ///  the `EventLoopFuture` returned by `task`.
+    ///
+    /// - parameters:
+    ///     - task: The synchronous task to run. As everything that runs on the `EventLoop`, it must not block.
+    /// - returns: An `EventLoopFuture` identical to the `EventLooopFuture` returned from `task`.
+    public func flatSubmit<T>(_ task: @escaping () throws -> EventLoopFuture<T>) -> EventLoopFuture<T> {
```
Comment (weissi) [inline Sources/NIO/EventLoop.swift]: the task closure returns a future so it shouldn't be throwing.

File: Tests/NIOTests/EventLoopTest.swift [hunk 3]
```diff
@@ -613,6 +613,40 @@ public final class EventLoopTest : XCTestCase {
         XCTAssertFalse(channel.isActive)
     }
 
+    public func testFlatSubmitCloseFutureNotifiedBeforeUnblock() throws {
+        class AssertHandler: ChannelInboundHandler {
```
Comment (weissi) [inline Tests/NIOTests/EventLoopTest.swift]: this test case tests the new `flatSubmit` functionality but also a quite a bit of unrelated other unrelated stuff. I think we could reduce this test case to only the things that are actually needed such as:

- a test case that `flatSubmit`s a successful computation to `EmbeddedEventLoop` 
- a test case that `flatSubmit`s a failing computation to `EmbeddedEventLoop`
- a test case that submits a computation to a `MultiThreadedEventLoopGroup` that is already shut down.

as an example for the first case:

```
func testSuccessfulFlatSubmit() {
    let el = EmbeddedEventLoop()
    XCTAssertNoThrow(XCTAssertEqual(1, try el.flatSubmit {
        el.makeSucceededFuture(1)
    }.wait()))
}
```

File: Sources/NIO/EventLoop.swift (hunk 1 above)
Comment (2bjake) [reply Sources/NIO/EventLoop.swift]: Sounds good. Removed.

File: Sources/NIO/EventLoop.swift (hunk 2 above)
Comment (2bjake) [reply Sources/NIO/EventLoop.swift]: Nice. Removed throws.

File: Tests/NIOTests/EventLoopTest.swift (hunk 3 above)
Comment (2bjake) [reply Tests/NIOTests/EventLoopTest.swift]: Agreed. Thanks for the example. I was unfamiliar with the EmbeddedEventLoop, so that was helpful.

**Comment by weissi** [comment]
@swift-nio-bot test this please

File: Tests/NIOTests/EventLoopTest.swift [hunk 4]
```diff
@@ -928,4 +928,48 @@ public final class EventLoopTest : XCTestCase {
...
+            XCTAssertEqual(.failed, error as? TestError)
+        }
+    }
+
+    func testFlatSubmitOnShutdownLoopGroup() {
+        let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        let eventLoop = eventLoopGroup.next()
+        XCTAssertNoThrow(try eventLoopGroup.syncShutdownGracefully())
+
+        let _ = eventLoop.flatSubmit { () -> EventLoopFuture<Void> in
```
Comment (weissi) [inline Tests/NIOTests/EventLoopTest.swift]: shame that our `submit` API doesn't return a failed future if the ELG is shut down but that's definitely not this PR's responsibility.

Comment (2bjake) [reply Tests/NIOTests/EventLoopTest.swift]: I had to experiment to determine what the expected behavior was. On my machine, the tests would hang if I called ```wait``` after submitting a task to a shutdown group.

However, I noticed that the tests seem to be failing intermittently on the build server (one failed twice, one failed only once). So it seems that the submitted tasks are somehow running after the group is shutdown? I haven't seen that behavior on my machine.

Comment (weissi) [reply Tests/NIOTests/EventLoopTest.swift]: --> 1177

Comment (2bjake) [reply Tests/NIOTests/EventLoopTest.swift]: Ok, sounds good.

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Pull request merged**